dotnet/sdk-phase-2.md

```diff
@@ -0,0 +1,484 @@
+        {
+            // Store current greeting
+            currentGreeting = await Workflow.ExecuteActivityAsync(
+                Activities.CreateGreeting,
+                await updates.ReceiveAsync(),
+                new(StartToCloseTimeout: TimeSpan.FromSeconds(5)));
+            Workflow.Logger.LogDebug("Greeting set to {Greeting}", currentGreeting);
+        }
+        return currentGreeting;
+    }
+
+    [WorkflowSignal]
+    public Task UpdateSalutationAsync(string salutation) =>
+        updates.SendAsync(new(Salutation: salutation, Name: name));
+
+    [WorkflowSignal]
+    public Task CompleteWithGreetingAsync()
+    {
+        updates.Complete();
+        return Task.CompletedTask;
+    }
+
+    [WorkflowQuery]
+    public string CurrentGreeting() => currentGreeting;
+}
+```
+
+Some things to not about the above code (much copied from phase 1):
+
+* Every workflow must have the non-inheritable `[Workflow]` attribute
+  * Optional `Name` positional property can be given. Default is the unqualified type name with the "I" prefix removed
+    if it's an interface and the next letter is also capital
+    * So what does `IDWorkflow` become? Yes, `DWorkflow` :-( Weighed as an acceptably rare cost for obvious benefit
+  * Non-inheritable means just because the interface or base class defines `[Workflow]`, the registered one must too
+* Workflows can have a single public constructor with `[WorkflowInit]` that take the exact same parameters as the
+  `[WorkflowRun]` method
+  * This is not required and most workflows probably won't use it but it is nice for those wanting to build readonly
+    properties instead of waiting until run
+  * If this is not present, a public parameterless constructor must be present (default of course in C# when no
+    constructors present)
+  * Ignored if in base class
+* Every workflow must explicitly have one and only one public method with a non-inheritable `WorkflowRun` attribute
+  * Can be in base class, but must also be overridden in this workflow class w/ the attribute even if it just delegates
+    to super
+  * Method must return a `Task`
+* Workflow signals are public async methods with the non-inheritable `[WorkflowSignal]` attribute
+  * Can be in base class, but if overridden, must also have the attribute
+  * Method must return a `Task` and can accept arguments
+    * It is a bit annoying to force Task on people that may not be doing an async thing but:
+      * They may change it later so might as well require it (especially since the "Async" suffix is usually present)
+      * Causes too many overloads if we accept `Task` and `void`
+  * Optional `Name` positional property can be given. Default is the unqualified method name with "Async" trimmed off if
+    present.
+* Workflow queries are public methods with the non-inheritable `[WorkflowQuery]` attribute
+  * Can be in base class, but if overridden, must also have the attribute
+  * Method cannot return a `Task`, but can accept arguments
```

Comment:
How hard would it be to add support for `Task` returning queries? I'm asking since we're talking about allowing starting local activities from update validators and queries.

Reply:
I agree with Chad. I don't think we should even call such things "local activities" if we allow it. We should just say "you can disable the sandbox and do stuff from queries if you really want, just don't mutate state, pinky swear". I don't think making it a "real" local activity really buys people anything, but does wildly complicate internals.

Reply:
I think the reason why we do want to leverage activities here is for dependency injection.
I don't know if we'll want to call the API local activity, I agree that is confusing.
I do see a reason to eventually support returning tasks and also using a mechanism supported by the worker so we can avoid triggering the deadlock detector or in generally blocking the workflow thread.
We don't need to design this now though.
I was just pointing out that this might need to change and wanted to understand how hard it would be to change when needed.